x/ibc/light-clients/07-tendermint/types/upgrade.go

```diff
@@ -26,102 +25,128 @@ import (
-	updatedClientState := NewClientState(
-		tmCommittedClient.ChainId, cs.TrustLevel, cs.TrustingPeriod, tmCommittedClient.UnbondingPeriod,
-		cs.MaxClockDrift, tmCommittedClient.LatestHeight, tmCommittedClient.ProofSpecs, tmCommittedClient.UpgradePath,
+	newClientState := NewClientState(
+		tmUpgradeClient.ChainId, cs.TrustLevel, cs.TrustingPeriod, tmUpgradeClient.UnbondingPeriod,
+		cs.MaxClockDrift, tmUpgradeClient.LatestHeight, tmUpgradeClient.ProofSpecs, tmUpgradeClient.UpgradePath,
 		cs.AllowUpdateAfterExpiry, cs.AllowUpdateAfterMisbehaviour,
 	)
 
-	if err := updatedClientState.Validate(); err != nil {
+	if err := newClientState.Validate(); err != nil {
 		return nil, nil, sdkerrors.Wrap(err, "updated client state failed basic validation")
 	}
 
-	if err := merkleProof.VerifyMembership(cs.ProofSpecs, consState.GetRoot(), upgradePath, bz); err != nil {
-		return nil, nil, err
-	}
+	// The new consensus state is merely used as a trusted kernel against which headers on the new
+	// chain can be verified. The root is empty as it cannot be known in advance, thus no proof verification will pass.
+	// The timestamp of the consensus state is also this chain's blocktime. This is because starting up a new chain
+	// may take a long time, especially if there are unexpected issues and thus we do not want the new client to be
+	// automatically expired due to unforeseen delays.
+	newConsState := NewConsensusState(
+		ctx.BlockTime(), commitmenttypes.MerkleRoot{}, tmUpgradeConsState.NextValidatorsHash,
+	)
 
-	// TODO: Return valid consensus state https://github.com/cosmos/cosmos-sdk/issues/7708
-	return updatedClientState, &ConsensusState{}, nil
+	return newClientState, newConsState, nil
 }
 
-// construct MerklePath from upgradePath
-func constructUpgradeMerklePath(upgradePath string, upgradeHeight exported.Height) (commitmenttypes.MerklePath, error) {
-	// assume that all keys here are separated by `/` and
-	// any `/` within a merkle key is correctly escaped
-	upgradeKeys := strings.Split(upgradePath, "/")
-	// unescape the last key so that we can append `/{height}` to the last key
-	lastKey, err := url.PathUnescape(upgradeKeys[len(upgradeKeys)-1])
-	if err != nil {
-		return commitmenttypes.MerklePath{}, err
-	}
-	// append upgradeHeight to last key in merkle path
+// construct MerklePath for the committed client from upgradePath
+func constructUpgradeClientMerklePath(upgradePath []string, lastHeight exported.Height) commitmenttypes.MerklePath {
+	// copy all elements from upgradePath except final element
+	clientPath := make([]string, len(upgradePath)-1)
+	copy(clientPath, upgradePath)
+
+	// append lastHeight and `upgradedClient` to last key of upgradePath and use as lastKey of clientPath
+	// this will create the IAVL key that is used to store client in upgrade store
+	clientPath = append(clientPath, fmt.Sprintf("%s/%d/%s", upgradePath[len(upgradePath)-1], lastHeight.GetVersionHeight(), upgradetypes.KeyUpgradedClient))
```

Comment:
nit: I think readability could be improved by breaking this up a little, it's hard to verify correctness

Reply:
agree